sensor: fix: decouple Sensor Async API request from its execution

ubieda opened the pull request:
### Description
This PR introduces a shim layer to decouple the Sensor Read Async API from its execution. It achieves this by introducing the RTIO Workqueues service, relying on P4WQ's, and a pool of pre-allocated work items to process the requests.

Fixes #73676

### Changes
- Add RTIO Workqueues service.
- Add testsuite to validate RTIO Workqueues service.
- Incorporate RTIO Work requests to all sensor drivers implementing the submit API (this is the bug-fix).

### Testing
- Sensors Generic Testsuite
``` console
twister -p native_sim --inline-logs -vvv -s tests/drivers/build_all/sensor/sensors.generic_test
```
- RTIO Workq testsuite
``` console
twister -p native_sim --inline-logs -vvv -s tests/subsys/rtio/workq/rtio.workq
```
- Shell sample
``` console
west build -b croxel_cx1825/nrf52840 samples/shell/shell_module -- -DCONFIG_SENSOR -- -DCONFIG_SENSOR_SHELL=y
```
- DHT Polling sample
``` console
west build -b <target> samples/sensor/dht_polling
```

ubieda committed:
sensor: Enable RTIO Semaphores for Async API

In order to decouple the read request from its implementation, we need
to be able to yield to potential lower-priority threads, which is not
possible with k_yield() itself. As such, using RTIO Consume semaphore
enables us to achieve this.

This patch also refactors blocking API (sensor_read) to also block upon
consumption, so as not to require RTIO_SUBMIT_SEM enabled.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda commented:
FYI - This bug also is also present in the sensing subsystem. However, this fix does not address that since it goes directly to the drivers submit implementation:

https://github.com/zephyrproject-rtos/zephyr/blob/02ad7ef1f9cd3ea455dedbf62fc3a5cb0d73025d/subsys/sensing/sensing_sensor.c#L16-L32

ubieda committed:
tests: sensor: generic: Set thread priorities for back-to-back readings

In order to support back-to-back readings, it's required that the
RTIO workq threads priority is higher than the caller issuing the
back-to-back requests, otherwise the requests will fail as P4WQ won't
have the opportunity to clear the is_done semaphore, and the subsequent
request will be rejected.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

teburd commented:
I like that its a small patch and likely does fix the issue. I do wonder though if we're going to be working around functional sensor drivers with this. How would this work with a correctly written submit call that *is* asynchronous as would be expected? How do we help guide sensor developers with this change to do the right thing?

My proposition here would be instead to create a sensor driver context that sensor driver authors can use to implement submit the right way in an opt in fashion.

It can do mostly the same thing but be internal to the driver rather than putting the shim up front and center as this does.

E.g.

``` c 
struct sensor_submit_p4wq_task {
   struct k_p4wq
   struct rtio_iodev_sqe *sqe;
};
```

Along with some helpers drivers can use to implement the submit API correctly with blocking API calls.

Does that make sense? I could be wrong!

ubieda commented:
@teburd Thanks for your feedback!

Overall, I don't have a strong opinion: I think both are valid and have their pros/cons. 

Some thoughts:

Since this is a temporary solution, my preference would be to:
- Avoid duplication as much as practical, and
- Make it easy to remove after it's no longer needed.

>  I do wonder though if we're going to be working around functional sensor drivers with this.
 
 That's true. With this existing approach we're applying the workaround for all implementations: _good and bad_. However, this may not be significant for the upcoming release as I don't think any of the existing in-tree implementations meet the criteria of "good example" (maybe I'm wrong and I'm missing out :') ).

> How do we help guide sensor developers with this change to do the right thing?

This one is a tough question, as hiding it from the common sensor path may also mean it's "more" exposed to the sensor drivers - e.g: developers may think that since existing in-tree `.submit` are implemented using the P4WQ + blocking-handler this is expected way to work and we start seeing more instances following that pattern.

Ideally, this would be some suggestions on how we could make this better:
- **Improve documentation**: of `sensor_submit_t` to indicate driver implementors these shall follow asynchronous/non-blocking pattern (and RTIO-based).
-  **Consolidate Reference Implementation**: we basically "fix" one instance to be completely non-blocking and utilize RTIO API calls (counting on #73869 coming up soon).

However, this doesn't seem like a feasible approach considering the stage of the upcoming release.

Having said all of that - let's go with whichever one we feel most comfortable with and let's focus on the path moving forward. I'm happy to take this PR either way: LMK if, after reading through these comments, you still lean towards your proposition and I'll spend some time thinking how to restructure it to avoid duplication and meet this goal at the same time.

teburd commented:
There's 4 drivers in the tree (5 if you include the fallback) that effectively block on submit. I don't think it'd be too much work to reuse p4wq with a helper to convert the async task.

Exposing the p4wq shim here as it is, only to delete it right after seems wrong. I'd rather us provide it as a proper tool, noting that it is a workaround/shim for drivers to use. I get the idea of reducing code duplication, I'm suggesting its still going to be pretty minimal (a small .submit redirect function and driver data member add and init).

I think we need some others to chime in here though to perhaps break the tie @MaureenHelm @yperess

ubieda commented:
Sounds reasonable - I'll work on refactoring it, thanks!

ubieda committed:
rtio: Add RTIO Work-queues service

Adds ability to process synchronous requests in an asynchronous
fashion, relying on dedicated P4WQ's and pre-allocated work items.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
sensor: fallback: Decouple RTIO request using RTIO workqueues service

Incorporate RTIO workqueues to turn sensor_submit_callback into an
asynchronous request.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
sensor: akm09918c: Use RTIO workq on akm09918c_submit

To make its execution path non-blocking from the caller's perspective.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
sensor: bma4xx: Use RTIO workq on bma4xx_submit

To make its execution path non-blocking from the caller's perspective.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
sensor: bme280: Use RTIO workq on bme280_submit

To make its execution path non-blocking from the caller's perspective.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
sensor: icm42688: Use RTIO workq on icm42688_submit

To make its execution path non-blocking from the caller's perspective.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

teburd commented:
This is looking super nice, I'd actually go ahead and simply call it rtio_p4wq rather than async_shim as its more widely useable than simply as a shim. There will likely always be blocking actions that may wish to be done asynchronously.

Great job!

ubieda committed:
tests: rtio: workq: Add testsuite to exercise RTIO work-queues

Basic tests demonstrating ability to decouple, batch-submit, preempt
and keep track of used items.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda commented:
> This is looking super nice, I'd actually go ahead and simply call it rtio_p4wq rather than rtio_async_shim as its more widely useable than simply as a shim. There will likely always be blocking actions that may wish to be done asynchronously. I know that's going back on what I said earlier about it being a shim, but seeing it here I think it is a proper tool now to solve a problem and should be named as such.
> 
> 
> 
> Basically, from how I've read io_uring this is more or less how io_uring works internally (thread pool).
> 
> 
> 
> Great job! +1 from me with/without rename, up to you on that one.

Thanks! I agree the way it is structured now will be much more useful than before (hopefully serve more use cases, e.g: foundation to implement bus drivers fallback).

The reason CI is red now is that I added a test suite to validate different scenarios, but I didn't run forgot to run it with asserts enabled. I'll come back to it tomorrow.

WRT the naming (rtio_async_shim vs rtio_p4wq): I don't love "async_shim", but I do find it intuitive serving the current use-case, whereas rtio_p4wq would not ring a bell initially (highly subjective and could be completely wrong). Perhaps we could brainstorm what problems we believe this is meant to solve? Ideally, I'd lean towards being intuitive on _what_ and not so much on _how_, but that may be just wishful thinking.

teburd commented on drivers/sensor/Kconfig:
```diff
@@ -24,6 +24,8 @@ config SENSOR_ASYNC_API
 	select RTIO
 	select RTIO_SYS_MEM_BLOCKS
+	select RTIO_CONSUME_SEM
+	select RTIO_WORKQ
```
These options should likely be done at a driver level, e.g. sensor/bme280/Kconfig should select it

ubieda replied:
That's true - In this case, though:
- `RTIO_CONSUME_SEM` is needed even with no `RTIO_WORKQ`  so that `sensor_processing_with_callback` does not block using k_yield(), which could be a problem depending on thread priorities.
- `RTIO_WORKQ` is needed for `sensor_submit_fallback`.

As an action item, I've added `RTIO_WORKQ` to each individual sensor driver so that when the sensor shim goes, it'll be up to each driver to scrub it on its own time.

teburd replied:
Fair points! It’s fine then

teburd commented on drivers/sensor/asahi_kasei/akm09918c/akm09918c_async.c:
```diff
@@ -1,15 +1,18 @@
  */
 
 #include <zephyr/logging/log.h>
+#include <zephyr/rtio/work.h>
```
Nice

teburd commented on subsys/rtio/rtio_workq.c:
```diff
@@ -0,0 +1,93 @@
+	const struct device *dev = req->dev;
+	struct rtio_iodev_sqe *iodev_sqe = req->iodev_sqe;
+
+	req->handler(dev, iodev_sqe);
```
The handler shouldn't need device, the iodev_sqe has a link to the iodev which should link to the device much like it does already in submit

ubieda replied:
You're right - I've scrubbed this from the APIs. Each handler should know how to retrieve its own device reference (e.g: sensors do it through `sensor_read_config`, SPI through `spi_dt_spec` and so forth).

Thanks!

teburd reviewed (commented):
Looking better with each push
